### DISCUSSION TITLE
Why are the input buffers of ahead-of-time compiled functions non-const

### DISCUSSION THREAD

**Discussion by dragly** [halide/Halide#7301] [General]
Looking at for instance the example in the tutorial, the input buffers of AOT-generated functions are non-const:

```c
int brighter(halide_buffer_t *_input_buffer, uint8_t _offset, halide_buffer_t *_brighter_buffer);
```

Is there a reason why these functions need to be able to modify the buffers? It would be nice to call the functions with a const buffer.

**Comment by abadams** [comment]
Halide supports bounds queries. If you set the host and device pointers to null, and pass in an output buffer of the shape you want, Halide will populate the input buffer fields to tell you how much input is required. This lets you tile Halide pipelines from the outside without knowing anything about the algorithm. It also means that sometimes input buffer fields will be mutated, so the input buffer can't be const.

**Reply by dragly** [reply]
Thanks for the detailed answer! I knew about bounds queries in the context of defining extern functions, but have not thought about that they could be used to query an AOT function as well. That makes a lot of sense and could actually become useful for us in some contexts :+1:
